### PR TITLE
feat: add parsing email addresses, phones, links from cv

### LINKED ISSUE #57: Parsing of email, phone number & links from uploaded CV files
Rabbit holes
Important to note that links in LinkedIn CVs are often split across two lines, so would be good to extract them correctly in such cases as well

### PR DESCRIPTION
**Pull request by PrinceYauj**
Closes #57.

- [x] I have reviewed my code in "Files changed" tab.
- [x] I have re-read the issue and this PR fully resolves it.

### CODE REVIEW

File: app/controllers/api/v1/documents_controller.rb [hunk 1]
```diff
@@ -16,8 +16,13 @@ def create
       case create_or_update_candidate(params_hash, url)
       in Success(candidate)
         case add_resume(candidate, file)
-        in Success(file)
-          render json: { url: candidate.url }, status: :ok
+        in Success()
+          case update_profile(candidate, file)
```
Comment (PrinceYauj) [inline app/controllers/api/v1/documents_controller.rb:20]: Need some cool idea on how to reduce the amount of nesting

File: app/controllers/ats/candidates_controller.rb [hunk 2]
```diff
@@ -432,24 +432,44 @@ def upload_file
...
+          cv_file: file
+        ).call
+        in Success()
+          nil
+        in Failure[:contacts_invalid, data]
+          warning = t("candidates.upload_cv.contacts_invalid", data:)
+        in Failure[:invalid_pdf, data]
+          warning = t("candidates.upload_cv.invalid_pdf", data:)
+        in Failure[:parse_pdf_error, data]
+          warning = t("candidates.upload_cv.parse_pdf_error", data:)
```
Comment (PrinceYauj) [inline app/controllers/ats/candidates_controller.rb:465]: I want to group these, but idk how to get failure's message (the only way I see is to pass it as an additional value)

File: app/controllers/api/v1/documents_controller.rb [hunk 3]
```diff
@@ -68,6 +73,14 @@ def add_resume(candidate, file)
     ).call
   end
 
+  def update_profile(candidate, file)
```
Comment (wrmk) [inline app/controllers/api/v1/documents_controller.rb:76]: i think we need to clarify name `update_profile_from_resume` smth like this

**Review by wrmk — COMMENTED** [review]
may we add tests for this logic?

did you try on [this file](https://github.com/toughbyte/ats/issues/338#issuecomment-2390639215)?

btw, maybe we make `Candidates::UpdateFromCV` nested for `Candidates::UploadFile` ?
and call if file is cv

File: app/controllers/api/v1/documents_controller.rb [hunk 4]
```diff
@@ -16,8 +16,13 @@ def create
...
       in Success(candidate)
         case add_resume(candidate, file)
-        in Success(file)
-          render json: { url: candidate.url }, status: :ok
+        in Success()
+          case update_profile(candidate, file)
+          in Success()
+            render json: { url: candidate.url }, status: :ok
+          in Failure[:contacts_invalid, :e]
+            render json: { url: candidate.url, message: error_message(e) }, status: :ok
```
[the remark this answers is not in the document]
Comment (PrinceYauj) [reply app/controllers/api/v1/documents_controller.rb:24]: You mean, we will create candidate anyway and if contacts parsing fails, we won't tell the user anything and send error to sentry, correct?

File: app/controllers/ats/candidates_controller.rb (hunk 2 above)
Comment (PrinceYauj) [reply app/controllers/ats/candidates_controller.rb:465]: actually, I just followed the usual workflow to handle every possible error. But I'm fine with sending it to sentry

File: app/operations/candidates/update_from_cv.rb [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+  end
+
+  def extract(text_to_parse, country_code:)
+    CVParser::Content.extract_from_text(text_to_parse, country_code:)
+  end
+
+  def update_contacts(data, parsed_emails:, parsed_urls:, country_code:, actor_account:, candidate:)
+    phones = (candidate.candidate_phones.map do |candidate_phone|
+                candidate_phone
+                  .slice(:phone, :list_index, :status, :source, :type, :created_via).symbolize_keys
```
[the remark this answers is not in the document]
Comment (PrinceYauj) [reply app/operations/candidates/update_from_cv.rb:45]: It calls `candidate.change` in the end, that allows string params, it seems. Here we call an operation

File: app/operations/candidates/update_from_cv.rb [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+              end +
+              data.phones.filter_map do |phone|
+                next unless CandidatePhone.valid_phone?(phone, country_code)
+
+                {
+                  phone: CandidatePhone.normalize(phone, country_code),
+                  status: "current",
+                  type: "personal"
```
[the remark this answers is not in the document]
Comment (PrinceYauj) [reply app/operations/candidates/update_from_cv.rb:53]: idk, didn't bother. Lemme dig into it

File: app/controllers/ats/candidates_controller.rb [hunk 7]
```diff
@@ -432,24 +432,44 @@ def upload_file
...
-      file: candidate_params[:file],
-      cv: true
-    ).call
-    in Success(file)
-      redirect_to tab_ats_candidate_path(@candidate, :info)
-    in Failure[:file_invalid, e]
-      render_error e, status: :unprocessable_entity
+    file = candidate_params[:file]
+    # TODO: remove transaction
+    ActiveRecord::Base.transaction do
```
[the remark this answers is not in the document]
Comment (PrinceYauj) [reply app/controllers/ats/candidates_controller.rb:445]: Yes, I've added it temporarily so that I don't need to remove cv manually every time

**Comment by PrinceYauj** [comment]
> did you try on [this file](https://github.com/toughbyte/ats/issues/338#issuecomment-2390639215)?

Yep, tried and it worked

File: app/controllers/api/v1/documents_controller.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (wrmk) [reply app/controllers/api/v1/documents_controller.rb:24]: yes

File: app/controllers/ats/candidates_controller.rb (hunk 2 above)
Comment (wrmk) [reply app/controllers/ats/candidates_controller.rb:465]: yeah, we handle, but display nothing to user, just for us in sentry

File: app/operations/candidates/update_from_cv.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (PrinceYauj) [reply app/operations/candidates/update_from_cv.rb:53]: Well, type is mandatory in `Candidates::Change` operation:
```ruby
phones?: Types::Strict::Array.of(
      Types::Strict::Hash.schema(
        phone: Types::Strict::String,
        status?: Types::Strict::String,
        source?: Types::Strict::String,
        type: Types::Strict::String
      ).optional
```

[the remark this answers is not in the document]
Comment (PrinceYauj) [reply app/operations/candidates/update_from_cv.rb:53]: btw, idk why is that. In our internal project the phone, parsed from cv has `type` nil.

[the remark this answers is not in the document]
Comment (wrmk) [reply app/operations/candidates/update_from_cv.rb:55]: looks like a mistake
 @Lyams do you remember why we allowed `type` as nil?

File: test/operations/candidates/apply_test.rb [hunk 8]
```diff
@@ -6,6 +6,7 @@ class Candidates::ApplyTest < ActionDispatch::IntegrationTest
   include Dry::Monads[:result]
 
   test "apply should create candidate, placement and task and assign recruiter" do
+    skip "doesnt work"
```
Comment (PrinceYauj) [inline test/operations/candidates/apply_test.rb:9]: I was unable to fix it. Tried ideas [from here](https://stackoverflow.com/questions/4511586/rails-functional-test-case-and-uploading-files-to-actiondispatchhttpuploadfi) but to no avail

File: app/operations/candidates/update_from_cv.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (wrmk) [reply app/operations/candidates/update_from_cv.rb:55]: cc @Lyams

File: app/controllers/ats/candidates_controller.rb [hunk 9]
```diff
@@ -432,24 +432,44 @@ def upload_file
...
+        in Success()
+          nil
+        in Failure[:contacts_invalid, data]
+          warning = t("candidates.upload_cv.contacts_invalid", data:)
+        in Failure[:invalid_pdf, data]
+          warning = t("candidates.upload_cv.invalid_pdf", data:)
+        in Failure[:parse_pdf_error, data]
+          warning = t("candidates.upload_cv.parse_pdf_error", data:)
+        in Failure(:unsupported_file_format)
+          warning = t("candidates.upload_cv.unsupported_format")
```
[the remark this answers is not in the document]
Comment (wrmk) [reply app/controllers/ats/candidates_controller.rb]: decided to not display any alert

File: app/operations/candidates/upload_file.rb [hunk 10]
```diff
@@ -14,6 +14,8 @@ def call
       add_event(attachment:, file:, actor_account:)
     end
 
+    update_profile_from_cv(candidate:, file:, actor_account:) if cv
```
Comment (wrmk) [inline app/operations/candidates/upload_file.rb:17]: please use `case` for this operation calling
just display that we do nothing

File: app/operations/candidates/update_from_cv.rb [hunk 11]
```diff
@@ -10,7 +10,8 @@ class Candidates::UpdateFromCV < ApplicationOperation
   def call
     country_code = candidate.location&.country_code
     file_extension = cv_file.original_filename.split(".").last
-    return Failure(:unsupported_file_format) if file_extension != "pdf"
+
+    return Failure() if file_extension != "pdf"
```
Comment (wrmk) [inline app/operations/candidates/update_from_cv.rb:14]: please clarify what failure here and below

File: app/operations/candidates/upload_file.rb (hunk 10 above)
Comment (PrinceYauj) [reply app/operations/candidates/upload_file.rb:17]: case here, or inside the method?

Comment (wrmk) [reply app/operations/candidates/upload_file.rb:17]: hmmm, i'd prefer here. but up to you

File: app/operations/candidates/update_from_cv.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (wrmk) [reply app/operations/candidates/update_from_cv.rb:55]: does not remember, let's assume that in freeats type is a must

**Review by wrmk — APPROVED** [review]

**Pull request merged**